File: src/main/java/yaasl/server/providers/Ktrax.java
package yaasl.server.providers;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.boot.json.JsonParser;
import org.springframework.boot.json.JsonParserFactory;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Component;
import org.springframework.web.client.RestTemplate;
import yaasl.server.model.Aircraft;
import yaasl.server.model.Flight;
import yaasl.server.model.Location;

import java.util.*;

import static java.lang.Integer.parseInt;
import static java.util.Calendar.MINUTE;
import static java.util.Collections.emptyList;
import static java.util.Comparator.comparing;
import static java.util.stream.Collectors.toList;
import static org.apache.commons.collections4.CollectionUtils.isNotEmpty;
import static org.apache.commons.lang3.time.DateUtils.*;
import static org.springframework.http.HttpStatus.OK;
import static yaasl.server.convert.Converter.formatDate;
import static yaasl.server.convert.Converter.parseDate;

@Component
public class Ktrax {

    private Logger LOG = LoggerFactory.getLogger(getClass());

    @Value("${provider.ktrax.url}")
    private String ktraxURL;

    private JsonParser jsonParser = JsonParserFactory.getJsonParser();

    public List<Flight> getFlights(String location, String date) {
        String url = ktraxURL + "?db=sortie&query_type=ap";
        if (location != null) {
            url += "&id=" + location.toUpperCase();
        }
        if (date != null) {
            Date dateBegin = parseDate(date);
            url += "&dbeg=" + formatDate(dateBegin) + "&dend=" + formatDate(dateBegin);
        }
        RestTemplate restTemplate = new RestTemplate();
        ResponseEntity<String> response = restTemplate.getForEntity(url, String.class);
        if (response.getStatusCode() == OK) {
            return process(response.getBody(), new Date());
        }
        else {
            LOG.error("Unable to query Ktrax (http code: {})", response.getStatusCode());
            return null;
        }
    }

    private List<Flight> process(String json, Date date) {
        Map<String, Object> data = jsonParser.parseMap(json);
        List<Map<String, Object>> sorties = (List<Map<String, Object>>) data.get("sorties");
        if (isNotEmpty(sorties)) {
            List<Flight> flights = new ArrayList<Flight>();
            Deque<Flight> allFlights = new ArrayDeque<Flight>(sorties.stream().map(sortie -> process(sortie, date)).collect(toList()));
            while (!allFlights.isEmpty()) {
                Flight flight = allFlights.pop();
                Flight matchingFlight = findMatchingFlight(flight, allFlights);
                if (matchingFlight != null) {
                    if (flight.getAircraft().isCanTow()) {
                        matchingFlight.setTowPlane(flight.getAircraft());
                        matchingFlight.setTowPlaneLandingTime(flight.getLandingTime());
                        flights.add(matchingFlight);
                    }
                    else {
                        flight.setTowPlane(matchingFlight.getAircraft());
                        flight.setTowPlaneLandingTime(matchingFlight.getLandingTime());
                        flights.add(flight);
                    }
                    allFlights.remove(matchingFlight);
                }
                else {
                    flights.add(flight);
                }
            }
            flights.sort(comparing(Flight::getStartTime));
            return flights;
        }
        else {
            return emptyList();
        }
    }

    private Flight findMatchingFlight(Flight flight, Deque<Flight> allFlights) {
        String towIdentifier = flight.getAircraft().getTowIdentifier();
        if (towIdentifier != null) {
            Optional<Flight> matchingFlight = allFlights
                    .stream()
                    .filter(possibleMatchingFlight -> towIdentifier.equals(possibleMatchingFlight.getAircraft().getIdentifier()))
                    .findFirst();
            return matchingFlight.isPresent() ? matchingFlight.get() : null;
        }
        return null;
    }

    private Flight process(Map<String, Object> sortie, Date date) {
        Flight flight = new Flight();

        int type = (sortie.get("type") != null ? (Integer) sortie.get("type") : 10);
        Aircraft aircraft = new Aircraft();
        aircraft.setCallSign((String) sortie.get("cs"));
        aircraft.setCompetitionNumber(!"-".equals(sortie.get("cn")) ? (String) sortie.get("cn") : null);
        aircraft.setCanTow(type == 2);
        aircraft.setNeedsTowing(type == 1);
        aircraft.setIdentifier((String) sortie.get("id"));
        aircraft.setTowIdentifier((String) sortie.get("tow_id"));
        flight.setAircraft(aircraft);

        Map<String, Object> tkof = (Map<String, Object>) sortie.get("tkof");
        flight.setStartTime(prepareDate((String) tkof.get("time"), date));
        Location startLocation = new Location();
        startLocation.setName((String) tkof.get("loc"));
        flight.setStartLocation(startLocation);

        Map<String, Object> ldg = (Map<String, Object>) sortie.get("ldg");
        flight.setLandingTime(prepareDate((String) ldg.get("time"), date));
        Location landingLocation = new Location();
        landingLocation.setName((String) ldg.get("loc"));
        flight.setLandingLocation(landingLocation);

        return flight;
    }

    private Date prepareDate(String time, Date date) {
        String[] hourAndMinutes = time.split(":");
        if (hourAndMinutes.length == 2) {
            return truncate(setMinutes(setHours(date, parseInt(hourAndMinutes[0])), parseInt(hourAndMinutes[1])), MINUTE);
        }
        else {
            return null;
        }
    }

}

File: src/main/kotlin/yaasl/server/providers/Ktrax.kt
package yaasl.server.providers

import org.apache.commons.collections4.CollectionUtils.isNotEmpty
import org.apache.commons.lang3.time.DateUtils.*
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.boot.json.JsonParserFactory
import org.springframework.http.HttpStatus.OK
import org.springframework.stereotype.Component
import org.springframework.web.client.RestTemplate
import yaasl.server.convert.Converter.formatDate
import yaasl.server.model.Aircraft
import yaasl.server.model.Flight
import yaasl.server.model.Location
import java.lang.Integer.parseInt
import java.util.*
import java.util.Calendar.MINUTE
import java.util.Collections.emptyList

@Component
class Ktrax(@Value("\${provider.ktrax.url}") private val ktraxURL: String) {

    private val LOG = LoggerFactory.getLogger(javaClass)
    private val jsonParser = JsonParserFactory.getJsonParser()

    fun getFlights(location: Location?, date: Date?): List<Flight> {
        var url = ktraxURL + "?db=sortie&query_type=ap"
        if (location != null) {
            url += "&id=" + location.icao
        }
        if (date != null) {
            url += "&dbeg=" + formatDate(date) + "&dend=" + formatDate(date)
        }
        val restTemplate = RestTemplate()
        val response = restTemplate.getForEntity<String>(url, String::class.java)
        if (response.statusCode == OK) {
            return process(response.body, Date())
        } else {
            LOG.error("Unable to query Ktrax (http code: {})", response.statusCode)
            return emptyList()
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun process(json: String?, date: Date): List<Flight> {
        val data = jsonParser.parseMap(json)
        val sorties = data["sorties"] as List<Map<String, Any>>
        if (isNotEmpty(sorties)) {
            val flights = ArrayList<Flight>()
            val allFlights = ArrayDeque(sorties.map { sortie -> process(sortie, date) }.toList())
            while (!allFlights.isEmpty()) {
                val flight = allFlights.pop()
                val matchingFlight = findMatchingFlight(flight, allFlights)
                if (matchingFlight != null) {
                    if (flight.aircraft!!.canTow) {
                        matchingFlight.towPlane = flight.aircraft
                        matchingFlight.towPlaneLandingTime = flight.landingTime
                        flights.add(matchingFlight)
                    } else {
                        flight.towPlane = matchingFlight.aircraft
                        flight.towPlaneLandingTime = matchingFlight.landingTime
                        flights.add(flight)
                    }
                    allFlights.remove(matchingFlight)
                } else {
                    flights.add(flight)
                }
            }
            flights.sortBy { it.startTime }
            return flights
        } else {
            return emptyList()
        }
    }

    private fun findMatchingFlight(flight: Flight, allFlights: Deque<Flight>): Flight? {
        val towIdentifier = flight.aircraft!!.towIdentifier
        if (towIdentifier != null) {
            return allFlights
                    .filter { possibleMatchingFlight -> towIdentifier == possibleMatchingFlight.aircraft!!.identifier }
                    .firstOrNull()
        }
        return null
    }

    @Suppress("UNCHECKED_CAST")
    private fun process(sortie: Map<String, Any>, date: Date): Flight {
        val flight = Flight()

        flight.sequence = sortie["seq"] as Long

        val type = if (sortie["type"] != null) sortie["type"] as Int else 10
        val aircraft = Aircraft()
        aircraft.callSign = sortie["cs"] as String
        aircraft.competitionNumber = if ("-" != sortie["cn"]) sortie["cn"] as String else null
        aircraft.canTow = type == 2
        aircraft.needsTowing = type == 1
        aircraft.identifier = sortie["id"] as String
        aircraft.towIdentifier = sortie["tow_id"] as String
        flight.aircraft = aircraft

        val tkof = sortie["tkof"] as Map<String, Any>
        flight.startTime = prepareDate(tkof["time"] as String, date)
        val startLocation = Location()
        startLocation.icao = tkof["loc"] as String
        flight.startLocation = startLocation

        val ldg = sortie["ldg"] as Map<String, Any>
        flight.landingTime = prepareDate(ldg["time"] as String, date)
        val landingLocation = Location()
        landingLocation.icao = ldg["loc"] as String
        flight.landingLocation = landingLocation

        return flight
    }

    private fun prepareDate(time: String, date: Date): Date? {
        val hourAndMinutes = time.split(":".toRegex()).dropLastWhile({ it.isEmpty() }).toTypedArray()
        return if (hourAndMinutes.size == 2) {
            truncate(setMinutes(setHours(date, parseInt(hourAndMinutes[0])), parseInt(hourAndMinutes[1])), MINUTE)
        } else {
            null
        }
    }

}
